From c1eb1ba9e4b9d4b5c8eba18509c52db970c24874 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Wed, 7 Dec 2016 13:54:44 +0200
Subject: [PATCH 0645/1345] fix: net: mvpp2x: fix two crashes with all ports
 disabled state

commit  85e8322fe3d8d4e61c273cf66891d8059067622e from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

1. First crash:
- Driver will return -ENODEV if port count is 0 and there are no
  enabled ports in device tree.
2. Second crash:
- create single thread workqueue in the later states of driver probe.
  Crash occurred since workqueue were created and driver exit probe
  with error. To solve this issue workqueue creation were moved to the end
  of the probe.

Change-Id: I50006c6ef07c0872bc7566a16a44ebf1545614a7
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/34445
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c |   10 +---------
 1 file changed, 1 insertion(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index aa2d319..395bbfa 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -5146,7 +5146,7 @@ static int mv_pp2x_platform_data_get(struct platform_device *pdev,
 	*port_count = of_get_available_child_count(dn);
 	if (*port_count == 0) {
 		dev_err(&pdev->dev, "no ports enabled\n");
-		err = -ENODEV;
+		return -ENODEV;
 	}
 	return 0;
 }
@@ -5329,14 +5329,6 @@ static int mv_pp2x_probe(struct platform_device *pdev)
 		goto err_clk;
 	}
 
-	priv->workqueue = create_singlethread_workqueue("mv_pp2x");
-
-	if (!priv->workqueue) {
-		err = -ENOMEM;
-		goto err_clk;
-	}
-	INIT_DELAYED_WORK(&priv->stats_task, mv_pp2x_get_device_stats);
-
 	/* Init PP22 rxfhindir table evenly in probe */
 	if (priv->pp2_version == PPV22)
 		mv_pp22_init_rxfhindir(priv);
-- 
1.7.9.5

